src/main/java/com/arpnetworking/metrics/mad/sources/DimensionInjectingSource.java

```diff
@@ -0,0 +1,195 @@
+            final Map<String, String> newDimensions;
+            if (_replace) {
+                newDimensions = Maps .newHashMapWithExpectedSize(record.getAnnotations().size() + _dimensions.size());
+                newDimensions.putAll(record.getDimensions());
+                newDimensions.putAll(_dimensions);
+            } else {
+                newDimensions = Maps .newHashMapWithExpectedSize(record.getAnnotations().size() + _dimensions.size());
+                newDimensions.putAll(_dimensions);
+                newDimensions.putAll(record.getDimensions());
+            }
+
+            _source.notify(ThreadLocalBuilder.build(
+                    DefaultRecord.Builder.class,
+                    b1 -> b1.setMetrics(record.getMetrics())
+                            .setId(record.getId())
+                            .setTime(record.getTime())
+                            .setAnnotations(record.getAnnotations())
+                            .setDimensions(ImmutableMap.copyOf(newDimensions))));
+        }
+
+        private final DimensionInjectingSource _source;
+        private final boolean _replace;
+        private final ImmutableMap<String, String> _dimensions;
+    }
+
+    /**
+     * Implementation of builder pattern for {@link DimensionInjectingSource}.
+     *
+     * @author Brandon Arp (brandon dot arp at inscopemetrics dot com)
+     */
+    public static final class Builder extends BaseSource.Builder<Builder, DimensionInjectingSource> {
+
+        /**
+         * Public constructor.
+         */
+        public Builder() {
+            super(DimensionInjectingSource::new);
+        }
+
+        /**
+         * Sets the underlying source. Required. Cannot be null.
+         *
+         * @param value The underlying source.
+         * @return This instance of {@link Builder}.
+         */
+        public Builder setSource(final Source value) {
+            _source = value;
+            return this;
+        }
+
+        /**
+         * Sets the dimensions. Required. Cannot be null.
+         *
+         * @param value The map of dimensions and their values
+         * @return This instance of {@link Builder}.
+         */
+        public Builder setDimensions(final Map<String, String> value) {
+            _dimensions = value;
+            return this;
+        }
+
+        /**
+         * Sets whether or not to replace existing values. Optional. Cannot be null. Defaults to true.
+         *
+         * @param value True to replace values
+         * @return This instance of {@link Builder}.
+         */
+        public Builder setReplaceExisting(final Boolean value) {
+            _replaceExisting = value;
+            return this;
+        }
+
+        @Override
+        protected Builder self() {
+            return this;
+        }
+
+        @NotNull
+        private Source _source;
+        @NotNull
+        private Map<String, String> _dimensions;
+        @NotNull
+        private Boolean _replaceExisting = true;
```

Comment:
Should this be a per dimension setting? I guess you could nest two of these sources; however, that is more data copy. After the work on the mapping dimensions, I've been thinking that we could do this type of injection closer to the source and therefore avoid the copying altogether.

Reply:
I'd rather not complicate it like that. We haven't even had this source at all. Mixing id's probably unnecessary for now. If there is a use case I can add it.

Reply:
Sure we haven't had this because most of the use case have not supported arbitrary dimension aggregation; we've had the equivalent sink though.